### PR TITLE
Allow to read jobs in Scheduler

### PR DESCRIPTION
**Pull request by rmanibus**
This is a first step to implement #23632 by allowing to read jobs in the scheduler.

### CODE REVIEW

**Comment by rmanibus** [comment]
While I'm on it, is it on purpose ?
https://github.com/quarkusio/quarkus/blob/1cca7d9d11a55cd1173bb793bcd8e7c067332bf1/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L504

**Comment by rmanibus** [comment]
The TriggerStateTest for Quartz is failing because Quartz Scheduler does not set the TriggerState to Error when a job throw an exception.
I tried to override the method
```
public CompletedExecutionInstruction executionComplete(JobExecutionContext context,
                                                               JobExecutionException result)
``` 
in the quartz trigger to return SET_TRIGGER_ERROR if there is an exception, but it seems that when the trigger is in error state, it is not scheduled anymore.

Maybe something can be done with a SchedulerListener, or maybe storing the error state in the JobDataMap but it is not persistent at the moment, it would require to enable @PersistJobDataAfterExecution. I dont know how it could be done otherwise in clustered mode as we do not have a full view of the jobs on one node.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (rmanibus) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:343]: Maybe this need to be package private ?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: Could pls describe what each state constant actually means?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:343]: +1. In fact, the same applies to `isRunning()`.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: Yes I made a comment about that, for now it's mirroring Quartz TriggerState, but we could create something more explicit. I was thinking of: 
RUNNING, 
PAUSED,
LAST_EXECUTION_ERROR: Exception during last execution
ERROR: Scheduler error
OVERDUE: would include BLOCKED status from Quartz, and locally computed overdue state based on next execution time

**Comment by mkouba** [comment]
> While I'm on it, is this done on purpose ?
> 
> https://github.com/quarkusio/quarkus/blob/1cca7d9d11a55cd1173bb793bcd8e7c067332bf1/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L502-L510

That's definitely a bug ;-). However, I think that it should be fixed in a separate PR so that it can be easily backported. Would you care to send one?

**Comment by rmanibus** [comment]
> > While I'm on it, is this done on purpose ?
> > https://github.com/quarkusio/quarkus/blob/1cca7d9d11a55cd1173bb793bcd8e7c067332bf1/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L502-L510
> 
> That's definitely a bug ;-). However, I think that it should be fixed in a separate PR so that it can be easily backported. Would you care to send one?

yes I'm on it !

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java]: done

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: TBH I'm not convinced this would be superuseful. `RUNNING/PAUSED` basically copies the functionality of `Scheduler.isRunning()` and `Scheduler.isPaused(String)`. `LAST_EXECUTION_ERROR/ERROR` only covers the last "result" and in fact, an exception thrown from a job does not always have to be an error. And `OVERDUE` is disputable because in theory you can "compute" the overdue value from the  `Trigger.getNextFireTime()`.

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: yes I agree there is some overlap between the status and the methods from the scheduler. I still think we should be able to now if there was some exception thrown in a task. taking only the last one allow at least to spot a problem if it's happening on every run, but maybe we can do better (emitting a cdi event and let the user decide what to do ?).

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: @mkouba I think the isOverdue method is useful because next fire time can be null, and this method can check the delay.
but to be more consistant with the rest of the api, it might be better to place it in the scheduler rather than in the Trigger, what do you think ?
```
        @Override
        public Instant getNextFireTime() {
            Date nextFireTime = context.getTrigger().getNextFireTime();
            Date nextFireTime = trigger.getNextFireTime();
            return nextFireTime != null ? nextFireTime.toInstant() : null;
```

Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: > . Ok, so isOverdue() + configurable OVERDUE_GRACE_PERIOD_IN_SECONDS seems to be useful.

+1, if we can make things configurable. But I've a more basic questions for me to be able to try and fully understand what we are trying to do here: 

@rmanibus @mkouba what is our intentions with these states and how do we expect the user to use them?

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:14]: @machi1990 I will remove the state, the goal of this is just to be able see what task are in the scheduler, and to be able to evaluate the health of each job.

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java]: @mkouba @machi1990 I modified the PR according to this discussion, can you have a look ?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (machi1990) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]:
```suggestion
    @ConfigItem(defaultValue = "1")
```

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (machi1990) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:22]: Should this be `overdueGracePeriod`?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java]: I see thanks for the explanation and the PR. 

I've small code suggestion and renaming of the configuration. I'll let @mkouba have a general overview. 

Can we also update the documentation on the feature?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:22]: In the label also or just the variable name ?

Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:22]: I'd suggest that the user see `quarkus.scheduler.overdue-grace-period`.

If you change the variable name, the configuration name will be automatically generated from it.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: The name is not needed since it can be generated from the variable name.

If the config is of type Duraton and it is just a number, it is automatically considered as seconds.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java]: ok thanks, it's done !

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java]: it's done

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java]: @machi1990 thank you I updated the PR according to the suggestions. I will have a look at the doc

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java]: @machi1990 I updated the documentation, tell me if it's ok for you !

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/FailedExecution.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/FailedExecution.java:11]: Pls remove these fields. We kept them on `SkippedExecution` just for backwards compatibility. `ScheduledExecution` holds this information as well.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/SuccessExecution.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/SuccessExecution.java:11]: The same applies here ;-) https://github.com/quarkusio/quarkus/pull/23639/files#r807743630

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/SuccessExecution.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/SuccessExecution.java:8]: It should be `SuccessfulExecution` to keep the names consistent ;-)

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java:33]: It would not hurt to mention that the period is configurable. WDYT?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:19]: And the unit is second, or?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: 1 second is IMO a bit too strict default. Actually, now when I think about it... it might make more sense to configure the overdue per job, i.e. add something like `@Scheduled#overduePeriod()`. @machi1990 WDYT?

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:19]: well, it's a Duration, so the unit is defined by the user

Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:19]: Well, a user does not use the config class directly. Very often the `application.properties` file is used. 

You're right. The processing of a duration format is standardized in quarkus (and mentioned in any config reference guide).

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java]: I changed it for: @return true if the last execution of the trigger is overdue. The grace period is configurable 

let me know if it's enough !

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: Hi  @mkouba, updated the PR according to your review. should I implement this one also ?

Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: I'd like to hear from @machi1990 about his opinion ;-)

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java
Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Trigger.java]: It depends ;-). If we move the config to the `@Scheduled#overduePeriod()` then a simple link would be enough I guess.

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java
Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: I can see the usefulness of both options. Having a global one like the way we do here will enable for an overall control of the overdue for all the schedulers: it is like a default overdue duration. @mkouba you were thinking of having one or the other?

Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: We can have both and `@Scheduled#overduePeriod()` defaulted to the global one...

Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: > We can have both and `@Scheduled#overduePeriod()` defaulted to the global one...

+1, @rmanibus what do you think about this?

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: Yes, it will obviously be more flexible, I will try to update the PR today and add some tests for this. For Quartz, as we do not have a global object representing the job, should I rather add the grace period in the jobdata as a primitive, or create a map storing the durations (like for the invokers) ? What will happen if I store it as a map and two instances of a clustered scheduler run a different version of the app ?

Comment (mkouba) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: > For Quartz, as we do not have a global object representing the job...

Hm, maybe it would make more sense to create a global map, i.e. identity -> job metadata.

Comment (machi1990) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SchedulerRuntimeConfig.java:21]: > > For Quartz, as we do not have a global object representing the job...
> 
> Hm, maybe it would make more sense to create a global map, i.e. identity -> job metadata.

+1

**Review by machi1990 — APPROVED** [review]
lgtm, thanks for the great contribution @rmanibus 

Leaving it to @mkouba to get his approval too. 

I am also thinking maybe it will be good to get some code examples in our quickstarts. But let's get this in first.

**Comment by rmanibus** [comment]
> ## Failing Jobs - Building [2242012](https://github.com/quarkusio/quarkus/commit/2242012d0f9b8bc690cc4056c67c91b44a3aab1d)
> Status 	Name 	Step 	Failures 	Logs 	Raw logs
> ✖ 	JVM Tests - JDK 11 	`Build` 	[Failures](#user-content-test-failures-job-5235026982) 	[Logs](https://github.com/quarkusio/quarkus/runs/5235026982?check_suite_focus=true) 	[Raw logs](https://github.com/quarkusio/quarkus/commit/2242012d0f9b8bc690cc4056c67c91b44a3aab1d/checks/5235026982/logs)
> ✖ 	JVM Tests - JDK 11 Windows 	`Build` 	[Failures](#user-content-test-failures-job-5235027202) 	[Logs](https://github.com/quarkusio/quarkus/runs/5235027202?check_suite_focus=true) 	[Raw logs](https://github.com/quarkusio/quarkus/commit/2242012d0f9b8bc690cc4056c67c91b44a3aab1d/checks/5235027202/logs)
> ✔️ 	JVM Tests - JDK 17 				
> 
> Full information is available in the [Build summary check run](https://github.com/quarkusio/quarkus/runs/5235210948).
> ## Failures
> ### ⚙️ JVM Tests - JDK 11 [#](#user-content-test-failures-job-5235026982)
> 
> ```diff
> - Failing: extensions/scheduler/deployment 
> ! Skipped: extensions/quartz/deployment extensions/spring-scheduled/deployment integration-tests/devmode and 3 more
> ```
> 
> #### 📦 extensions/scheduler/deployment
> 
> ✖ `io.quarkus.scheduler.test.OverdueExecutionTest.testExecution` line `38` - [More details](https://github.com/quarkusio/quarkus/runs/5235210948#user-content-test-failure-io.quarkus.scheduler.test.overdueexecutiontest-1) - [Source on GitHub](https://github.com/rmanibus/quarkus/blob/2242012d0f9b8bc690cc4056c67c91b44a3aab1d/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/OverdueExecutionTest.java#L38)
> ### ⚙️ JVM Tests - JDK 11 Windows [#](#user-content-test-failures-job-5235027202)
> 
> ```diff
> - Failing: extensions/scheduler/deployment 
> ! Skipped: extensions/quartz/deployment extensions/spring-scheduled/deployment integration-tests/devmode and 3 more
> ```
> 
> #### 📦 extensions/scheduler/deployment
> 
> ✖ `io.quarkus.scheduler.test.OverdueExecutionTest.testExecution` line `38` - [More details](https://github.com/quarkusio/quarkus/runs/5235210948#user-content-test-failure-io.quarkus.scheduler.test.overdueexecutiontest-1) - [Source on GitHub](https://github.com/rmanibus/quarkus/blob/2242012d0f9b8bc690cc4056c67c91b44a3aab1d/extensions/scheduler/deployment/src/test/java/io/quarkus/scheduler/test/OverdueExecutionTest.java#L38)

There is a race condition in this test because the scheduler is paused before the first execution (or not) and in this case lastFireTime is null

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java:141]: Do we have a test for this?

**Review by mkouba — COMMENTED** [review]
Looks really good! I've added few minor comments.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/FailedExecutionTest.java
Comment (rmanibus) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/FailedExecutionTest.java:25]: I wanted to be sure it does not crash after first round but it might be tested else where !

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java:141]: well, it's the same parser as delay & every. 
in OverdueExecutionTest I use the simplified syntax (1s):
https://github.com/quarkusio/quarkus/blob/f95fc935b8b00b1f6e1ce3f33d7711241e42c3b4/extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/OverdueExecutionTest.java#L49-L51

Do you want a more specific test ?

Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/Scheduled.java:141]: I added a test to be sure

**Comment by rmanibus** [comment]
@mkouba, I made the changes and squashed everything.

File: extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/FailedExecutionTest.java
Comment (mkouba) [reply extensions/quartz/deployment/src/test/java/io/quarkus/quartz/test/FailedExecutionTest.java:25]: Fair enough.

File: docs/src/main/asciidoc/scheduler-reference.adoc
Comment (mkouba) [inline docs/src/main/asciidoc/scheduler-reference.adoc:284]: `scheduler.listScheduledJobs()` -> `scheduler.getScheduledJobs()`

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (mkouba) [inline extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:266]: The result is not immutable. So something like `return List.copyOf(identities)` or  `return scheduledTasks.stream().map(s -> s.trigger).collect(Collectors.toUnmodifiableList())`?

File: extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java
Comment (mkouba) [inline extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java:350]: The result is not immutable. So something like `return List.copyOf(scheduledTasks.values())`?

**Review by mkouba — COMMENTED** [review]
Thanks! One last series of comments :-)

File: extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java
Comment (rmanibus) [reply extensions/scheduler/runtime/src/main/java/io/quarkus/scheduler/runtime/SimpleScheduler.java:266]: When I looked at the code base, I haven't seen any stream so I thought it was not in the coding standard !

**Comment by rmanibus** [comment]
@mkouba it's done !

**Review by mkouba — APPROVED** [review]

**Pull request merged**
